### PR TITLE
Driver infrastructure improvements

### PR DESCRIPTION
**Pull request by mohrcore**
The DEVCLASS commit #725 introduced a problem where the drivers started to attach to wrong devices due to incorrent probing mechanism. Additionally, `device_t`'s  created in device detection mechanisms were not scanned when probing drivers for a bus device.

This issue solves #774

### CODE REVIEW

File: include/sys/pci.h [hunk 1]
```diff
@@ -114,4 +114,12 @@ static inline pci_device_t *pci_device_of(device_t *device) {
...
 }
 
+#define DEVICE_DRIVER_GEN_PCI_PROBE(vendor__id, device__id, device_name)       \
+  static int dev_generic_probe_##device_name(device_t *dev) {                  \
+    pci_device_t *pcid = pci_device_of(dev);                                   \
+    if (pcid->vendor_id != (vendor__id) || pcid->device_id != (device__id))    \
+      return 0;                                                                \
+    return 1;                                                                  \
+  }
+
```
Comment (pj1031999) [inline include/sys/pci.h:124]: Please don't define functions in that way. It is not acceptable.

`bus_generic_probe` is already in `sys/bus.h`. What is the use case for `dev_generic_probe_*`? Why it is defined here?

Please add comments.

File: sys/drv/ns16550.c [hunk 2]
```diff
@@ -183,11 +186,12 @@ static int ns16550_attach(device_t *dev) {
...
-  .attach = ns16550_attach,
-  .identify = bus_generic_identify,
-};
+DEVICE_DRIVER_GEN_PCI_PROBE(NS16550_VENDOR_ID, NS16550_DEVICE_ID, ns16550)
+
+static driver_t ns16550_driver = {.desc = "NS16550 UART driver",
+                                  .size = sizeof(ns16550_state_t),
+                                  .attach = ns16550_attach,
+                                  .identify = bus_generic_identify,
+                                  .probe = dev_generic_probe_ns16550};
```
Comment (pj1031999) [inline sys/drv/ns16550.c:195]: Why did you change the formatting here?

File: sys/drv/pit.c [hunk 3]
```diff
@@ -137,11 +140,13 @@ static int pit_attach(device_t *dev) {
...
-  .identify = bus_generic_identify,
-};
+DEVICE_DRIVER_GEN_PCI_PROBE(INTEL_PIIX4_IDE_VENDOR_ID,
+                            INTEL_PIIX4_IDE_DEVICE_ID, pit)
+
+static driver_t pit_driver = {.desc = "i8254 PIT driver",
+                              .size = sizeof(pit_state_t),
+                              .attach = pit_attach,
+                              .identify = bus_generic_identify,
+                              .probe = dev_generic_probe_pit};
```
Comment (pj1031999) [inline sys/drv/pit.c:150]: Why did you change the formatting here?

File: sys/drv/rtc.c [hunk 4]
```diff
@@ -126,11 +129,12 @@ static int rtc_attach(device_t *dev) {
...
-  .attach = rtc_attach,
-  .identify = bus_generic_identify,
-};
+DEVICE_DRIVER_GEN_PCI_PROBE(RTC_VENDOR_ID, RTC_DEVICE_ID, rtc)
+
+static driver_t rtc_driver = {.desc = "MC146818 RTC driver",
+                              .size = sizeof(rtc_state_t),
+                              .attach = rtc_attach,
+                              .identify = bus_generic_identify,
+                              .probe = dev_generic_probe_rtc};
```
Comment (pj1031999) [inline sys/drv/rtc.c:138]: Why did you change the formatting here?

File: sys/mips/gt64120.c [hunk 5]
```diff
@@ -396,13 +396,18 @@ static device_t *gt_pci_identify(driver_t *drv, device_t *parent) {
...
+}
+
 /* clang-format off */
 pci_bus_driver_t gt_pci_bus = {
   .driver = {
     .desc = "GT-64120 PCI bus driver",
     .size = sizeof(gt_pci_state_t),
     .attach = gt_pci_attach,
     .identify = gt_pci_identify,
+    .probe = gt64120_probe
```
Comment (pj1031999) [inline sys/mips/gt64120.c:410]: Why `gt64120_probe`? All other functions here are `gt_pci_*`.

File: sys/kern/bus.c [hunk 6]
```diff
@@ -87,6 +87,8 @@ void generic_bs_write_region_4(bus_space_handle_t handle, bus_size_t offset,
     *dst++ = *src++;
 }
 
+#define DEV_IDENTIFIED_BY_PARENT ((void *)0xffffffff)
```
Comment (pj1031999) [inline sys/kern/bus.c:90]: Any reason for this address? I'm afraid this could easily cause an error.
Why not `NULL`?

File: sys/kern/bus.c [hunk 7]
```diff
@@ -105,6 +107,18 @@ bus_space_t *generic_bus_space = &(bus_space_t){
...
+                          int *error) {
+  dev->driver = drv;
+  if (device_probe(dev)) {
+    klog("%s detected!", drv->desc);
+    *error = device_attach(dev);
+    if (*error)
+      return 1;
+  }
+  return 0;
+}
```
Comment (pj1031999) [inline sys/kern/bus.c:120]: Where are you using `bus` here?

I think that `error` is not needed here and we could unified `error` with return value.

**Review by pj1031999 — CHANGES_REQUESTED** [review]
Please fix code style and simplify code.

**Comment by cahirwpz** [comment]
@mohrcore @ThomasStachovsky Please respond to @pj1031999 remarks.

File: sys/drv/ns16550.c (hunk 2 above)
Comment (ThomasStachovsky) [reply sys/drv/ns16550.c]: I misued make format, it's corrected now.

File: sys/drv/pit.c (hunk 3 above)
Comment (ThomasStachovsky) [reply sys/drv/pit.c]: I misued make format, it's corrected now.

File: sys/drv/rtc.c (hunk 4 above)
Comment (ThomasStachovsky) [reply sys/drv/rtc.c]: I misued make format, it's corrected now.

File: sys/mips/gt64120.c (hunk 5 above)
Comment (ThomasStachovsky) [reply sys/mips/gt64120.c]: Changed to gt_pci_probe.

File: sys/kern/bus.c (hunk 6 above)
Comment (ThomasStachovsky) [reply sys/kern/bus.c:90]: It cannot be ```NULL``` without further changes to ```bus_generic_probe```. In ```bus_generic_probe```there is this code:
```
if (dev == NULL)
      continue;
if (dev == DEV_IDENTIFIED_BY_PARENT) {...}
``` 
I suppose the reason for this address being ```0xffffffff``` is that without changes in the logic of  ```bus_generic_probe``` this value cannot be ```NULL```, and there can't be a device at ```0xffffffff```.

File: sys/kern/bus.c (hunk 7 above)
Comment (ThomasStachovsky) [reply sys/kern/bus.c]: ```bus``` was unused and has been removed, ```error``` also has been removed from the parameters of the function.

File: sys/kern/bus.c (hunk 6 above)
Comment (pj1031999) [reply sys/kern/bus.c:90]: In that case in my opinion `-1` is better -- more portable -- value. Please remember that we support _AArch64_ where pointers have 64 bits.

File: include/sys/pci.h (hunk 1 above)
Comment (mohrcore) [reply include/sys/pci.h]: `DEVICE_DRIVER_GEN_PCI_PROBE` was used to generated generic probe methods that checked VENDOR_ID and DEVICE_ID fields of a pci device.

I just replaced it with handwritten code with more consistent naming conventions.

**Review by pj1031999 — APPROVED** [review]
LGTM
